### ISSUE TITLE
How do I make storeon/localstorage work with storeon/router?

### ISSUE THREAD

**Issue by Hubro** [storeon/localstorage#19]
I followed the example in the README, and my state was successfully stored to localStorage, but it was cleared every time I reloaded the window. After placing some breakpoints, it seems like the router module (`@storeon/router`) dispatches an event immediately when the page loaded, at which time the state was empty, and that caused `@storeon/localstorage` to overwrite the local storage with `undefined`.

Moving `persistState` to be the first module in `createStore` sort of fixed the problem, since it's now the first module to execute its `@init` handler. The problem is that the `@init` handler of my other modules will now overwrite the loaded state, so I have to modify them like this:

```typescript
store.on("@init", ({ auth }) => {
    return { auth: auth || { status: "anonymous" } }
});
```

Do you have any ideas how to make `@storeon/localstorage` work with `@storeon/router`?

**Comment by ai** [comment]
Why do you want to put `router` data to the `localstorage`?

**Comment by Hubro** [comment]
> Why do you want to put `router` data to the `localstorage`?

I don't, I just want to use both modules in the same application.

Currently my state has 2 keys, `router.key` and `"auth"`. I only persist the auth key:

```typescript
export interface IState {
    // Module: router
    [router.key]: any;

    // Module: auth
    auth: {
        status: "anonymous" | "logged-in";
        username?: string;
    }
}

const store = createStore<IState, IEvents>([
    routerModule,
    authModule,

    // Save the auth state to localstorage
    persistState<IState>(["auth"], { key: "my-app-state" }),
]);
```

When my application loads, the router dispatches an event during `@init`. This is from `@storeon/router/index.js`:

```javascript
    store.on('@init', function () {
      store.dispatch(change, parse(loc.pathname, routes))
    })
```

`@storeon/localstorage` listens for the `@dispatch` event and saves `state["auth"]` to local storage. This happens before the `@init` handler in `@storeon/localstorage` has been executes, so it hasn't loaded anything from local storage yet. `state["auth"]` therefore resolves to `undefined`, so it ends up overwriting the local storage with `undefined`.

To summarize, the local storage module overwrites the local storage with `undefined` before it loads it, and it happens because the router dispatches an event during `@init`, before the local storage module has processed its `@init` handler.

**Comment by ai** [comment]
@polemius let's add boolean variable `wasInited` and save store to localStorage only after `@init` event

**Comment by Hubro** [comment]
@ai I was thinking the same :) I made a pull request. I tested it and it solves the problem.

**Comment by ai** [comment]
Let’s wait for the maintainer (it is a night in Russia now).

**Comment by polemius** [comment]
I've merge #20 PR and released new 0.6.0 version. @Hubro please check if every thing works with your code.

**Comment by Hubro** [comment]
@polemius Installed latest version from npm now, and everything works perfectly :)
